ruby/ql/lib/codeql/ruby/ast/internal/Constant.qll

```diff
@@ -1,36 +1,464 @@
-  } or
-  TString(string s) { any(RequiredConstantValue x).requiredString(s) } or
-  TSymbol(string s) { any(RequiredConstantValue x).requiredSymbol(s) } or
-  TBoolean(boolean b) { b in [false, true] } or
-  TNil()
-
-private newtype TRequiredConstantValue = MkRequiredConstantValue()
+private import codeql.ruby.AST
+private import codeql.ruby.ast.internal.Literal
+private import codeql.ruby.ast.internal.Module
+private import codeql.ruby.controlflow.CfgNodes
+private import codeql.ruby.dataflow.SSA
+private import ExprNodes
 
 /**
- * A class that exists for QL technical reasons only (the IPA type used
- * to represent constant values needs to be bounded).
+ * Provides an implementation of constant propagation for control-flow graph
+ * (CFG) nodes and expressions (AST nodes).
+ *
+ * The end result are two predicates
+ * ```ql
+ * ConstantValue getConstantValue(ExprCfgNode n);
+ * ConstantValue getConstantValueExpr(Expr e);
+ * ```
+ *
+ * It would be natural to define those predicates recursively. However, because
+ * of how `newtype`s work, this results in bad performance as a result of
+ * unnecessary recursion through the constructors of `TConstantValue`. Instead,
+ * we define a set of predicats for each possible `ConstantValue` type, and each
+ * set of predicates need to replicate logic, e.g., how a constant may be propagated
+ * from an assignment to a variable read.
+ *
+ * For each `ConstantValue` type `T`, we define three predicates:
+ * ```ql
+ * predicate isT(ExprCfgNode n, T v);
+ * predicate isTExprNoCfg(Expr e, T v);
+ * predicate isTExpr(Expr e, T v);
+ * ```
+ *
+ * `isT` and `isTExpr` rely on the CFG to determine the constant value of a CFG
+ * node and expression, respectively, whereas `isTExprNoCfg` is able to determine
+ * the constant value of an expression without relying on the CFG. This means that
+ * even if the CFG is not available (dead code), we may still be able to infer a
+ * constant value in some cases.
  */
-class RequiredConstantValue extends MkRequiredConstantValue {
-  string toString() { none() }
+private module Propagation {
+  private ExprCfgNode getSource(VariableReadAccessCfgNode read) {
+    exists(Ssa::WriteDefinition def |
+      def.assigns(result) and
+      read = def.getARead()
+    )
+  }
+
+  predicate isInt(ExprCfgNode e, int i) {
+    isIntExprNoCfg(e.getExpr(), i)
+    or
+    isIntExpr(e.getExpr().(ConstantReadAccess).getValue(), i)
+    or
+    isInt(getSource(e), i)
+    or
+    e =
+      any(UnaryOperationCfgNode unop |
+        unop.getOperator() = "-" and
+        isInt(unop.getOperand(), -i)
+        or
+        unop.getOperator() = "+" and
+        isInt(unop.getOperand(), i)
+      )
+    or
+    exists(BinaryOperationCfgNode binop, string op, int left, int right |
+      e = binop and
+      isInt(binop.getLeftOperand(), left) and
+      isInt(binop.getRightOperand(), right) and
+      op = binop.getOperator()
+    |
+      op = "+" and
```

Comment:
Would it make any/much difference if we we avoided using `getOperator` and a string comparison, and instead wrote something like this?
```suggestion
      binop.getExpr() instanceof AddExpr and
```

Reply:
That is indeed a bit nicer, I will update.